### PR TITLE
Adding Feature base classes

### PR DESCRIPTION
**Pull request by patxibocos**

### CODE REVIEW

File: src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt [hunk 1]
```diff
@@ -39,16 +42,9 @@ class FeaturesGeneratorTest : BehaviorSpec({
...
-                        this[0].type shouldBe Optimizely::class.java
-                        this[1].type shouldBe String::class.java
-                    }
-                    with(featuresClientClass.methods.first()) {
-                        this.name shouldBe "isFeatureEnabled"
-                        this.parameters.first().type.toString() shouldBe "class what.ever.pack.age.Features"
-                    }
+                        compilationResult.classLoader.loadClass("$packageName.TestFeaturesClient")
+                    featuresClientClass.constructors.first().newInstance(mockk<Optimizely>(), "")
+                        .shouldBeInstanceOf<FeaturesClient<*>>()
```
Comment (ralphpina) [inline src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt]: I see this being able to init the client as a plus 👍. But I don't think this benefit is enough to merit the extra complexity.

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt [hunk 2]
```diff
@@ -43,68 +49,52 @@ private fun String.optimizelyFeatureKeyToEnumConstant(): String =
 
 /**
  * This will produce something that looks like this:
- * enum class Features(val key: String) {
- *    NEW_CHECKOUT_PAGE("new_checkout_page"),
+ * enum class Features : Feature {
+ *     NEW_CHECKOUT_PAGE {
+ *         override val key: String = "new_checkout_page"
+ *     },
```
Comment (ralphpina) [inline src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: From a user's perspective, the API will the same: `Features.NEW_CHECKOUT_PAGE.key`. 👍.

I know the code is generated, so how it looks bears little weight. But personally, I like this:
```kotlin
enum class Features(val key: String) {
    NEW_CHECKOUT_PAGE("new_checkout_page")
}
```
...more than this:
```kotlin
enum class Features : Feature {
    NEW_CHECKOUT_PAGE {
        override val key: String = "new_checkout_page"
     }
}
```

I know there is a benefit for tests. Are there other benefits to this change?

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt [hunk 3]
```diff
@@ -43,68 +49,52 @@ private fun String.optimizelyFeatureKeyToEnumConstant(): String =
...
  */
-private fun featuresClientTypeSpec(packageName: String): TypeSpec =
-    TypeSpec.classBuilder("FeaturesClient")
+private fun featuresClientTypeSpec(featuresEnumTypeName: TypeName): TypeSpec {
+    val featuresClientClazz = FeaturesClient::class.java
+    val featuresClientClassName =
+        ClassName(featuresClientClazz.`package`.name, featuresClientClazz.simpleName).parameterizedBy(
+            featuresEnumTypeName
+        )
+    return TypeSpec.classBuilder("TestFeaturesClient")
```
Comment (ralphpina) [inline src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: Is this a bug? You're generating a class called `TestFeaturesClient`, which would be available in production.

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt [hunk 4]
```diff
@@ -43,68 +49,52 @@ private fun String.optimizelyFeatureKeyToEnumConstant(): String =
...
- *        optimizely.isFeatureEnabled(feature.key, userId)
- * }
+ * class TestFeaturesClient(
+ *     optimizely: Optimizely,
+ *     userId: String
+ * ) : FeaturesClient<Features>(optimizely, userId)
  */
-private fun featuresClientTypeSpec(packageName: String): TypeSpec =
-    TypeSpec.classBuilder("FeaturesClient")
+private fun featuresClientTypeSpec(featuresEnumTypeName: TypeName): TypeSpec {
```
Comment (ralphpina) [inline src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: I don't see any benefit to generating this class.
```kotlin
class TestFeaturesClient(
  optimizely: Optimizely,
  userId: String
) : FeaturesClient<Features>(optimizely, userId)
```

Why not remove the `abstract` from `FeaturesClient` and always use that directly?

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt (hunk 2 above)
Comment (patxibocos) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: I agree that I prefer the first one which is more "natural".

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt (hunk 3 above)
Comment (patxibocos) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: I added a new commit renaming every Base class to be prefixed with `Base` 🙂

Comment (patxibocos) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: So the final client is again called `FeaturesClient`

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt (hunk 4 above)
Comment (patxibocos) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: Because of the generic type. Enum doesn't exist until the generator runs.

**Comment by patxibocos** [comment]
I understand all you say @ralphpina 
What I like for making it more type safe is because all the interfaces and abstract classes define a contract of how the implementation would be 🙂
Probably the base classes would be accesible from the clients (as the generated ones are extending/implementing them) but I don't think this would be a problem

**Comment by patxibocos** [comment]
I need to feel less lazy about documenting 😂 
Thanks a lot for documenting this PR with a such detailed comment 👏

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt (hunk 4 above)
Comment (ralphpina) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: I guess I'm not seeing the value of making it a generic. Apart from the testing example, are there other use cases that require the clients to be parameterized?

Comment (patxibocos) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: Maybe if for any reason we would like to provide the ability to get multiple clients? Not sure if anyone could be depending on different Optimizely projects for example...
The main reason for introducing this PR was for consistency with the existing implementation with how experiments are done 🙂 
But I completely agree that the parent class does not introduce any value but the easiness for us to change the "template" for generating the code.

**Comment by ralphpina** [comment]
> What I like for making it more type safe is because all the interfaces and abstract classes define a contract of how the implementation would be 🙂

That's fair 👍. And having a contract as you describe is useful. 

But, what use cases do we have which require this contract? Are you planning to add some functionality where this polymorphism is needed? 

Right now the use cases we have are to map Optimizely experiments/features into enums to make them discoverable and easy to check. Maybe I'm missing others?

> Probably the base classes would be accesible from the clients (as the generated ones are extending/implementing them) but I don't think this would be a problem

I don't think there's a big issue here 👍. But it would pollute the namespace, since these would show up in autocomplete. Another small issue is that it might make the API harder to grok, since there are these base classes.

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt (hunk 4 above)
Comment (ralphpina) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: > Maybe if for any reason we would like to provide the ability to get multiple clients? Not sure if anyone could be depending on different Optimizely projects for example...

You will always have new use cases that pop up in the future. Given enough time, some use cases will no longer be useful and therefore should be deprecated. Would it be an over-optimization to try to satisfy future use cases now?

For example, using multiple Optimizely projects will probably require more than just base classes for polymorphism. We would need to pass two tokens, etc.

> The main reason for introducing this PR was for consistency with the existing implementation with how experiments are done

Consistency is good 😁. Won't argue that.

> But I completely agree that the parent class does not introduce any value but the easiness for us to change the "template" for generating the code.

Could the "template" be the code in the `*Generator.kt` files?

**Comment by patxibocos** [comment]
I am closing this one as your feedback @ralphpina 🙂 
In the end there is no perfect solution, and balancing the two approaches I see the that the base classes are not bringing any value to the clients

**Pull request closed without merging**